singularity-eos/eos/eos_powermg.hpp

```diff
@@ -0,0 +1,440 @@
+  void _CheckPowerMG();
+  void _InitializePowerMG(const Real *expcoeffs);
+  PORTABLE_INLINE_FUNCTION Real _HugPressureFromDensity(const Real rho) const;
+  PORTABLE_INLINE_FUNCTION Real _HugTemperatureFromDensity(const Real rho) const;
+  PORTABLE_INLINE_FUNCTION Real _SN2Mp2(const Real x) const;
+  PORTABLE_INLINE_FUNCTION Real
+  _compBulkModulusFromDensityTemperature(const Real rho, const Real temp) const;
+  PORTABLE_INLINE_FUNCTION Real
+  _compBulkModulusFromDensityInternalEnergy(const Real rho, const Real sie) const;
+};
+
+inline void PowerMG::_CheckPowerMG() {
+
+  if (_rho0 < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required PowerMG model parameter rho0 < 0");
+  }
+  if (_T0 < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required PowerMG model parameter T0 < 0");
+  }
+  if (_Cv0 < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required PowerMG model parameter Cv0 < 0");
+  }
+  if (_K0toK40[0] < 0.0) {
+    PORTABLE_ALWAYS_THROW_OR_ABORT("Required PowerMG model parameter K0 < 0");
+  }
+  if (_Pmin >= 0.0) {
+    _Pmin = -1000 * _K0toK40[0];
+#ifndef NDEBUG
+    PORTABLE_WARN(
+        "PowerMG model parameter Pmin not set or positive. Reset to default (-1000*K0)");
+#endif // NDEBUG
+  }
+}
+
+inline void PowerMG::_InitializePowerMG(const Real *K0toK40input) {
+
+  // The PressureCoeffsK0toK40Size (=41) allowed K0 to K40 coefficients
+  // for the pressure reference curve vs rho
+  // are seldom all used so did not want to crowd the argument list with them.
+  // Instead I ask the host code to send me a pointer to this array so that I can
+  // copy it here. Not used coeffs should be set to 0.0 (of course).
+  for (int ind = 0; ind < PressureCoeffsK0toK40Size; ind++) {
+    _K0toK40[ind] = K0toK40input[ind];
+    if (_K0toK40[ind] != 0.0) _M = ind;
+  }
+}
+
+PORTABLE_INLINE_FUNCTION Real PowerMG::_HugPressureFromDensity(Real rho) const {
+  Real eta = 1.0 - robust::ratio(_rho0, rho);
+  Real value = 0.0;
+  for (int ind = _M; ind >= 1; ind--) {
+    value = eta * value + _K0toK40[ind];
+  }
+  value = _K0toK40[0] * eta * (eta * value + 1.0);
+  return value;
+}
+PORTABLE_INLINE_FUNCTION Real PowerMG::_HugTemperatureFromDensity(Real rho) const {
+  Real eta = 1.0 - robust::ratio(_rho0, rho);
+  Real gamma2 = _SN2Mp2(_G0 * eta);
+  Real sum = _M * _K0toK40[_M] * gamma2;
+  for (int ind = _M - 1; ind >= 1; ind--) {
+    gamma2 = _G0 * eta / (ind + 2) * gamma2 + 1.0 / (ind + 2);
+    sum = eta * sum + ind * _K0toK40[ind] * gamma2;
+  }
+  Real temp = _K0toK40[0] / _Cv0 / 2.0 / _rho0 * eta * eta * eta * sum;
+  temp = _T0 * exp(eta * _G0) + temp;
+  return temp;
+}
+PORTABLE_INLINE_FUNCTION Real PowerMG::_SN2Mp2(const Real x) const {
+  Real ind = _M + 2;
+  int maxind = 200;
+  Real ak = 1.0 / ind;
+  Real sum = ak;
+  Real pf = exp(x) - 1.0;
+  Real ek = pf * ak;
+  while ((ek > sum * 1.e-15) && (ind < maxind)) {
+    ind = ind + 1;
+    ak = ak * x / ind;
+    ek = pf * ak;
+    sum = sum + ak;
+  }
```

Comment:
I'm not sure I understand what's going on with this function... Is this constructing Taylor series coefficients?

Reply:
OK. This is described in Robinson's paper. It is to make it fast and accurate.
ek is the maximum of the reminder of the infinite sum so we cut the sum off when that is less that machine precision from the sum itself.

Reply:
This is the only major calculation for the temperature. The rest of the coifs are calculated from this via recursion relations. Really cool.

Reply:
Could you add a comment above the loop to the effect of what you said here? This sort of clever code behavior isn't obvious so I think it's useful for future developers if a comment indicates what the purpose is.

Reply:
Done